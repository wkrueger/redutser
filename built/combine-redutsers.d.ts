import { Reducer, Redutser, RedutserShort, ReducerDict } from "./redutser";
export declare type ReplaceReducerState<Red, UpperState> = Red extends Reducer<any, infer P> ? Reducer<UpperState, P> : never;
export declare type ReplaceDictState<Dict extends ReducerDict<any>, StateOuter> = {
    [DictKey in keyof Dict]: ReplaceReducerState<Dict[DictKey], StateOuter>;
};
/**
 * Applies "mapToKey" to a key-value pair (ReducerDict), returning a new Dict.
 */
export declare const liftDictState: <StateOuter>() => <K extends keyof StateOuter, Dict extends ReducerDict<StateOuter[K]>>(mapKey: K, dict: Dict) => ReplaceDictState<Dict, StateOuter>;
export declare type DictOfRedutser<R> = R extends Redutser<any, infer D> ? D : never;
export declare type LiftRedutserState<State, Red extends Redutser<any, any> | undefined> = Redutser<State, ReplaceDictState<DictOfRedutser<Red>, State>>;
/**
 * "Lifts up" the state of a redutser.
 */
export declare const liftRedutserState: <OuterState, K extends keyof OuterState, R extends RedutserShort<OuterState[K], any>>(state: OuterState, key: K, _red: R) => {
    creators: import("./redutser").ActionCreatorsFromReducerDict<ReplaceDictState<DictOfRedutser<R>, OuterState>>;
    reducer: (state: OuterState | undefined, action: ReturnType<import("./redutser").ActionCreatorsFromReducerDict<ReplaceDictState<DictOfRedutser<R>, OuterState>>[keyof DictOfRedutser<R>]>) => OuterState;
    initialState: OuterState;
    actionTypes: ReturnType<import("./redutser").ActionCreatorsFromReducerDict<ReplaceDictState<DictOfRedutser<R>, OuterState>>[keyof DictOfRedutser<R>]>;
    plug: () => {
        ownProps: <OwnProps = {}>() => {
            mapProps: <StateProps = OuterState, DispatchProps = {
                dispatch: Red_Dispatcher<ReturnType<import("./redutser").ActionCreatorsFromReducerDict<ReplaceDictState<DictOfRedutser<R>, OuterState>>[keyof DictOfRedutser<R>]>, OuterState>;
            }>(stateMapper?: (state: OuterState) => StateProps, dispatchMapper?: (dispatcher: Red_Dispatcher<ReturnType<import("./redutser").ActionCreatorsFromReducerDict<ReplaceDictState<DictOfRedutser<R>, OuterState>>[keyof DictOfRedutser<R>]>, OuterState>) => DispatchProps) => {
                component: import("../type-helpers").ComponentEnhancer<StateProps & DispatchProps & OwnProps, OwnProps>;
            };
        };
        mapProps: <StateProps = OuterState, DispatchProps = {
            dispatch: Red_Dispatcher<ReturnType<import("./redutser").ActionCreatorsFromReducerDict<ReplaceDictState<DictOfRedutser<R>, OuterState>>[keyof DictOfRedutser<R>]>, OuterState>;
        }>(stateMapper?: (state: OuterState) => StateProps, dispatchMapper?: (dispatcher: Red_Dispatcher<ReturnType<import("./redutser").ActionCreatorsFromReducerDict<ReplaceDictState<DictOfRedutser<R>, OuterState>>[keyof DictOfRedutser<R>]>, OuterState>) => DispatchProps) => {
            component: import("../type-helpers").ComponentEnhancer<StateProps & DispatchProps & {}, {}>;
        };
    };
    plugShort: () => <OwnProps = {}>() => <StateProps = OuterState, DispatchProps = {
        dispatch: Red_Dispatcher<ReturnType<import("./redutser").ActionCreatorsFromReducerDict<ReplaceDictState<DictOfRedutser<R>, OuterState>>[keyof DictOfRedutser<R>]>, OuterState>;
    }>(stateMapper?: (state: OuterState) => StateProps, dispatchMapper?: (dispatcher: Red_Dispatcher<ReturnType<import("./redutser").ActionCreatorsFromReducerDict<ReplaceDictState<DictOfRedutser<R>, OuterState>>[keyof DictOfRedutser<R>]>, OuterState>) => DispatchProps) => import("../type-helpers").ComponentEnhancer<StateProps & DispatchProps & OwnProps, OwnProps>;
    __redutser__: boolean;
    _reducerDict: ReplaceDictState<DictOfRedutser<R>, OuterState>;
};
