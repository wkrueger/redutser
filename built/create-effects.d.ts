/**
 * I gave up on making declaration augment work on this, so
 * currently this is kept only for reference
 */
import { RedutserShort } from "./redutser";
declare global  {
    interface Red_EffectsType<A, S> {
    }
}
export declare function createEffects<Red extends RedutserShort<any, any>>(redutser: Red): <FX extends {
    [k: string]: (...i: any[]) => any;
}>(fx: FX) => FX;
