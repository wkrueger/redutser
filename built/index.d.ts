export { createRedutser, createRedutser2, Redutser } from "./redutser";
export { subdomain, combineRedutsers } from "./subdomain";
export { liftRedutserState, liftDictState } from "./combine-redutsers";
export * from "../type-helpers";
