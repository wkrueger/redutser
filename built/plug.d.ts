import { ComponentEnhancer } from "../type-helpers";
import { RedutserShort as Redutser } from "./redutser";
export declare const plugShort: <Red extends Redutser<any, any>>(redutser: Red) => <OwnProps = {}>() => <StateProps = Red["initialState"], DispatchProps = {
    dispatch: Redutser.DefaultDispatcher<Red["actionTypes"], Red["initialState"]>;
}>(stateMapper?: (state: Red["initialState"]) => StateProps, dispatchMapper?: (dispatcher: Redutser.DefaultDispatcher<Red["actionTypes"], Red["initialState"]>) => DispatchProps) => ComponentEnhancer<StateProps & DispatchProps & OwnProps, OwnProps>;
export declare const plug: <Red extends Redutser<any, any>>(redutser: Red) => {
    ownProps: <OwnProps = {}>() => {
        mapProps: <StateProps = Red["initialState"], DispatchProps = {
            dispatch: Redutser.DefaultDispatcher<Red["actionTypes"], Red["initialState"]>;
        }>(stateMapper?: (state: Red["initialState"]) => StateProps, dispatchMapper?: (dispatcher: Redutser.DefaultDispatcher<Red["actionTypes"], Red["initialState"]>) => DispatchProps) => {
            component: ComponentEnhancer<StateProps & DispatchProps & OwnProps, OwnProps>;
        };
    };
    mapProps: <StateProps = Red["initialState"], DispatchProps = {
        dispatch: Redutser.DefaultDispatcher<Red["actionTypes"], Red["initialState"]>;
    }>(stateMapper?: (state: Red["initialState"]) => StateProps, dispatchMapper?: (dispatcher: Redutser.DefaultDispatcher<Red["actionTypes"], Red["initialState"]>) => DispatchProps) => {
        component: ComponentEnhancer<StateProps & DispatchProps & {}, {}>;
    };
};
