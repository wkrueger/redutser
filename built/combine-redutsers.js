"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const redutser_1 = require("./redutser");
/**
 * An implementation of the above pair.
 */
const mapToKey = () => (key) => ({
    mapToInner: (s) => s[key],
    mapToOuter: (sin, sout) => (Object.assign({}, sout, { [key]: sin })),
});
/**
 * Gets a reducer.
 * Returns a new reducer with its state "lifted" from the original,
 * according to the mapper pair.
 */
const mapReducerState = () => (mapper) => {
    return (reducer) => {
        return (stateOuter, p) => mapper.mapToOuter(reducer(mapper.mapToInner(stateOuter), p), stateOuter);
    };
};
/**
 * Applies "mapToKey" to a key-value pair (ReducerDict), returning a new Dict.
 */
exports.liftDictState = () => (mapKey, dict) => {
    return Object.keys(dict).reduce((out, dictKey) => {
        const mapObj = mapToKey()(mapKey);
        return Object.assign({}, out, { [dictKey]: mapReducerState()(mapObj)(dict[dictKey]) });
    }, {});
};
/**
 * "Lifts up" the state of a redutser.
 */
exports.liftRedutserState = (state, key, _red) => {
    const dictMapped = exports.liftDictState()(key, 
    //this might feel a bit hacky but thats what worked better wth the inference atm
    //a previous less "hacky" approach brought inference issues.
    _red._reducerDict);
    return redutser_1.createRedutser(state, dictMapped);
};
