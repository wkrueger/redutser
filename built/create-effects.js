"use strict";
/**
 * I gave up on making declaration augment work on this, so
 * currently this is kept only for reference
 */
Object.defineProperty(exports, "__esModule", { value: true });
function createEffects(redutser) {
    return (fx) => fx;
}
exports.createEffects = createEffects;
