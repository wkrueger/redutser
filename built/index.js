export { createRedutser, createRedutser2 } from "./redutser";
export { subdomain, combineRedutsers } from "./subdomain";
export { liftRedutserState, liftDictState } from "./combine-redutsers";
export * from "./plug";
//export * from "./create-effects"
