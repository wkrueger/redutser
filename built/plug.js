"use strict";
exports.__esModule = true;
var react_redux_1 = require("react-redux");
exports.plugShort = function (redutser) { return function () { return function (stateMapper, dispatchMapper) {
    if (stateMapper === void 0) { stateMapper = function (state) { return state; }; }
    if (dispatchMapper === void 0) { dispatchMapper = function (dispatch) { return ({ dispatch: dispatch }); }; }
    return react_redux_1.connect(stateMapper, dispatchMapper);
}; }; };
/*
export const plug = <Red extends Redutser<any, any>>(redutser: Red) => {
  return {
    ownProps: <OwnProps = {}>() => {
      return {
        mapProps: <
          StateProps = Red["initialState"],
          DispatchProps = {
            dispatch: Redutser.DefaultDispatcher<
              Red["actionTypes"],
              Red["initialState"]
            >
          }
        >(
          stateMapper: (state: Red["initialState"]) => StateProps = state =>
            state,
          dispatchMapper: (
            dispatcher: Redutser.DefaultDispatcher<
              Red["actionTypes"],
              Red["initialState"]
            >
          ) => DispatchProps = dispatch => ({ dispatch } as any)
        ): {
          component: ComponentEnhancer<
            StateProps & DispatchProps & OwnProps,
            OwnProps
          >
        } => {
          const component = (connect as any)(stateMapper, dispatchMapper as any)
          return { component }
        },
      }
    },
  }
}
*/
exports.plug = function (redutser) {
    return {
        ownProps: ownProps(),
        mapProps: mapProps()
    };
};
var ownProps = function () { return function () {
    return {
        mapProps: mapProps()
    };
}; };
var mapProps = function () { return function (stateMapper, dispatchMapper) {
    if (stateMapper === void 0) { stateMapper = function (state) { return state; }; }
    if (dispatchMapper === void 0) { dispatchMapper = function (dispatch) { return ({ dispatch: dispatch }); }; }
    var component = react_redux_1.connect(stateMapper, dispatchMapper);
    return { component: component };
}; };
