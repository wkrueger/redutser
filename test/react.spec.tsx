import React from "react"
import { createRedutser } from "../built"
import { createStore } from "redux"
import { Provider } from "react-redux"
import renderer from "react-test-renderer"

function createSomething() {
  const initialState = { a: 1, b: "b" }

  const red = createRedutser(initialState, {
    increment: (state, count: number) => ({
      ...state,
      a: state.a + count,
    }),
    concat: (state, act: { text: string }) => ({
      ...state,
      b: state.b + act.text,
    }),
    doNothing: state => state,
  })

  return red
}

const red = createSomething()
const creators = red.creators

test("Plug 1", () => {
  const store = createStore(red.reducer)
  const Plugged = red
    .plug()
    .mapProps()
    .component(props => {
      return (
        <div onClick={() => props.dispatch(creators.doNothing({}))}>
          a: {props.a}
          b: {props.b}
        </div>
      )
    })

  const tree = renderer
    .create(
      <Provider store={store as any}>
        <Plugged />
      </Provider>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test("Plug 2", () => {
  const store = createStore(red.reducer)
  const Plugged = red
    .plug()
    .ownProps<{ myOwnProp: string }>()
    .mapProps(
      state => ({ b: state.b }),
      dispatch => ({
        increment: (n: number) => dispatch(creators.increment(n)),
      })
    )
    .component(props => {
      // expect error props.a
      return (
        <div onClick={() => props.increment(3)}>
          a: {props.b}
          ownProp: {props.myOwnProp}
        </div>
      )
    })
  const tree = renderer
    .create(
      <Provider store={store as any}>
        <Plugged myOwnProp="own" />
      </Provider>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test("Plug 3", () => {
  const store = createStore(red.reducer)
  const Plugged = red
    .plug()
    .mapProps(state => ({ a: state.a }))
    .component(props => {
      return (
        <div onClick={() => props.dispatch(red.creators.doNothing({}))}>
          a: {props.a}
        </div>
      )
    })
  const tree = renderer
    .create(
      <Provider store={store as any}>
        <Plugged />
      </Provider>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test("PlugShort 1", () => {
  const store = createStore(red.reducer)
  const Plugged = red.plugShort()<{ myOwnProp: string }>()(
    state => ({ b: state.b }),
    dispatch => ({
      increment: (n: number) => dispatch(creators.increment(n)),
    })
  )(props => {
    return (
      <div onClick={() => props.increment(3)}>
        a: {props.b}
        ownProp: {props.myOwnProp}
      </div>
    )
  })
  const tree = renderer
    .create(
      <Provider store={store as any}>
        <Plugged myOwnProp="own" />
      </Provider>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
