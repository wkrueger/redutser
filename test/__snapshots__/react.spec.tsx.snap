// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plug 1 1`] = `
<div
  onClick={[Function]}
>
  a: 
  1
  b: 
  b
</div>
`;

exports[`Plug 2 1`] = `
<div
  onClick={[Function]}
>
  a: 
  b
  ownProp: 
  own
</div>
`;

exports[`Plug 3 1`] = `
<div
  onClick={[Function]}
>
  a: 
  1
</div>
`;

exports[`PlugShort 1 1`] = `
<div
  onClick={[Function]}
>
  a: 
  b
  ownProp: 
  own
</div>
`;
